File: agriculture_management/hydroponic/report/chart___water_quality/chart___water_quality.py
# Copyright (c) 2025, aminos and contributors
# For license information, please see license.txt

import frappe


def execute(filters=None):
    """
    This script provides data specifically for the Water Quality dashboard chart.
    It safely handles empty filters to prevent errors on initial load.
    """
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    """Defines the columns. These must match the 'X Field' and 'Y Axis' fields in the Dashboard Chart."""
    return [
        {"label": "Timestamp", "fieldname": "Timestamp", "fieldtype": "Datetime"},
        {"label": "Hydroponic System",
            "fieldname": "hydroponic_system", "fieldtype": "Data"},
        {"label": "pH", "fieldname": "pH", "fieldtype": "Float"},
        {"label": "EC", "fieldname": "EC", "fieldtype": "Float"},
        {"label": "Temp C°", "fieldname": "Temp", "fieldtype": "Float"},
    ]

File: agriculture_management/hydroponic/report/chart___water_quality/chart___water_quality.js
// Copyright (c) 2025, aminos and contributors
// For license information, please see license.txt

frappe.query_reports["Chart - Water Quality"] = {
	"filters": [
		{
			"fieldname": "hydroponic_system",
			"label": __("Hydroponic System"),
			"fieldtype": "Link",
			"options": "Asset",
			"reqd": 1, // Keeping this not required to allow the dashboard to load

			/**
			 * THIS IS THE NEW, ENHANCED LOGIC.
			 * This function runs every time the user clicks the dropdown.
			 */
			"get_query": function() {
				// This function must return an object with a 'filters' key.
				return {
					"filters": {
						// Filter 1: Only show assets from the correct category.
						"asset_category": "Hydroponic Systems",

						// Filter 2: Only show assets where the status is NOT 'Cancelled'.
						// The ['!=', 'Value'] syntax means "not equal to".
						"status": ["!=", "Cancelled"]
					}
				};
			}
		}
	]
};